safekeeper/src/receive_wal.rs

```diff
@@ -32,6 +38,105 @@ use tracing::*;
 use utils::id::TenantTimelineId;
 use utils::lsn::Lsn;
 
+/// Registry of WalReceivers (compute connections). Timeline holds it (wrapped
+/// in Arc).
+pub struct WalReceivers {
+    mutex: Mutex<WalReceiversShared>,
+}
+
+/// Id under which walreceiver is registered in shmem.
+type WalReceiverId = usize;
+
+impl WalReceivers {
+    pub fn new() -> Arc<WalReceivers> {
+        Arc::new(WalReceivers {
+            mutex: Mutex::new(WalReceiversShared { slots: Vec::new() }),
+        })
+    }
+
+    /// Register new walreceiver. Returned guard provides access to the slot and
+    /// automatically deregisters in Drop.
+    pub fn register(self: &Arc<WalReceivers>) -> WalReceiverGuard {
+        let slots = &mut self.mutex.lock().slots;
```

Comment:
I didn't know you can take a mutable reference to a field under MutexGuard, without leaving it in the scope, how does it work? `slots` has a type `&mut Vec<Option<WalReceiverState>>` and I don't understand at which moment MutexGuard is dropped.

Reply:
I guess you are referring to different function, `get_slot` -- in `register` things are simple, guard is released at the end of func and no references are returned.

get_slot is interesting in this regard, yes. So `MappedMutexGuard` provides the magic -- it is a reference to a subfield embedding locking, i.e. when it disappears lock is released.

Reply:
No, I'm asking about this function :)

> guard is released at the end of func and no references are returned.

I can't understand how/where Rust compiler keeps this guard, because `slots` type didn't mention any guard. It looks like some kind of compiler feature, but I don't know how it's named.
<img width="886" alt="image" src="https://github.com/neondatabase/neon/assets/8150127/d9747494-8ae2-4557-a896-55e0c726fe42">

Reply:
Ah cool, I haven't realized that :)

Reply:
<img width="851" alt="image" src="https://github.com/neondatabase/neon/assets/8150127/bf08729e-d034-479a-912f-d9aabe28cc8d">

It's the right type, everything compiles after you specify it explicitly.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e79d40535e8833d35f7bb23fdd1f79b5

MIR showed that it implicitly creates `let _ MutexGuard<> = self.mutex.lock()` and keeps it until the end of scope, as expected.